solvers/bcd_celer.py

```diff
@@ -0,0 +1,363 @@
+
+
+@njit
+def dual_scaling_mtl(Theta, X, C, skip):
+    n_features = X.shape[1]
+    norm_XT_Theta = np.zeros(n_features)
+    nrm = 0
+
+    for j in C:
+        if skip[j]:
+            continue
+        Xj_Theta_nrm = norm(X[:, j] @ Theta, ord=2)
+        norm_XT_Theta[j] = Xj_Theta_nrm
+        if Xj_Theta_nrm > nrm:
+            nrm = Xj_Theta_nrm
+    return nrm, norm_XT_Theta
+
+
+@njit
+def create_ws_mtl(prune, W, prios, p0, t, screened, C, n_screened, ws_size):
+    n_features, n_tasks = W.shape
+
+    if t == 0:
+        ws_size = p0
+        for j in range(n_features):
+            for k in range(n_tasks):
+                if W[j, k]:
+                    prios[j] = -1
+                    break
+    else:
+        nnz = 0
+        if prune:
+            for j in range(n_features):
+                if W[j, 0] != 0:
+                    prios[j] = -1
+                    nnz += 1
+            ws_size = 2 * nnz
+        else:
+            for k in range(ws_size):
+                if not screened[C[k]]:
+                    prios[C[k]] = -1
+            ws_size = 2 * ws_size
+    ws_size = min(n_features - n_screened, ws_size)
+    return ws_size
+
+
+@njit
+def create_accel_pt(epoch, gap_freq, alpha, R, out, last_K_R, U, UtU,
+                    verbose):
+    K = U.shape[0] + 1
+
+    if epoch // gap_freq < K:
+        last_K_R[(epoch // gap_freq), :] = R
+    else:
+        for k in range(K - 1):
+            last_K_R[k, :] = last_K_R[k + 1, :]
+        last_K_R[K - 1, :] = R
+        for k in range(K - 1):
+            U[k] = last_K_R[k + 1].ravel() - last_K_R[k].ravel()
+
+        # double for loop but small: K**2/2
+        for k in range(K - 1):
+            for j in range(k, K - 1):
+                UtU[k, j] = np.dot(U[k], U[j])
+                UtU[j, k] = UtU[k, j]
+
+        try:
+            anderson = np.linalg.solve(UtU, np.ones(UtU.shape[0]))
+        except Exception:
+            # np.linalg.LinAlgError
+            # Numba only accepts Error/Exception inheriting from the generic
+            # Exception class
+            if verbose:
+                print("Singular matrix when computing accelerated point.")
+        else:
+            anderson /= np.sum(anderson)
+
+            out[:] = 0
+            for k in range(K - 1):
+                out += anderson[k] * last_K_R[k, :]
+            out /= alpha
+            # out now holds the extrapolated dual point
+
+
+def bcd_epoch(C, norms_X_block, X, R, alpha, W, inv_lc):
+    n_tasks = R.shape[1]
+    W_j_new = np.zeros((1, n_tasks))
+    dgemm = _get_dgemm()
+    alpha_lc = alpha * inv_lc
+
+    for j in C:
+        if norms_X_block[j] == 0.:
+            continue
+        idx = slice(j, j+1)
+        W_j = W[idx, :]
+        X_j = X[:, idx]
+
+        # W_j_old = W[j].copy()
+
+        # W_j_new = X_j.T @ R * inv_lc[j]
+        dgemm(alpha=inv_lc[j], beta=0.0, a=R.T, b=X_j, c=W_j_new.T,
+              overwrite_c=True)
+
+        if W_j[0, 0] != 0:
```

Comment:
beware that this is dangerous (the first entry could be 0 while the full line is not zero)